deep_quoridor/src/agents/alphazero/mcts.py

```diff
@@ -147,32 +145,38 @@ def select(self, node: Node) -> Node:
             node = node.select(self.visited_states)
         return node
 
-    def search(self, initial_game: Quoridor):
-        root = Node(initial_game, ucb_c=self.ucb_c)
-        new_nodes = []
-
+    def search_batch(self, initial_games: list[Quoridor]):
         if self.k is not None:
-            num_actions = np.sum(np.array(initial_game.get_action_mask()) == 1)
-            num_iterations = self.k * num_actions
+            num_iterations = [self.k * int(np.sum(np.array(game.get_action_mask()) == 1)) for game in initial_games]
         else:
             assert self.n is not None, "n must be specified if k is not"
-            num_iterations = self.n
+            num_iterations = [self.n for _ in initial_games]
+
+        max_iterations = max(num_iterations)
+
+        roots = [Node(g, ucb_c=self.ucb_c) for g in initial_games]
+        for iteration in range(max_iterations):
+            need_evaluation = []  # (root, node)
+            for game_idx, root in enumerate(roots):
+                # The number of iterations may be less if using mcts_k
+                if iteration >= num_iterations[game_idx]:
+                    continue
 
-        for _ in range(num_iterations):
-            # Traverse down the tree guided by maximum UCB until we find a node to expand
-            node = self.select(root)
+                # Traverse down the tree guided by maximum UCB until we find a node to expand
+                node = self.select(root)
 
-            if node.game.is_game_over():
-                # The player who just made a move must have won.
-                node.backpropagate_result(1)
-            elif self.max_steps >= 0 and node.game.completed_steps >= self.max_steps:
-                node.backpropagate_result(0)
-            else:
-                games_to_evaluate = [n.game for n in new_nodes]
-                new_nodes = []
+                if node.game.is_game_over():
+                    # The player who just made a move must have won.
+                    node.backpropagate_result(1)
+                elif self.max_steps >= 0 and node.game.completed_steps >= self.max_steps:
+                    node.backpropagate_result(0)
+                else:
+                    need_evaluation.append((root, node))
```

Comment:
Instead of evaluating the games, we gather what we need and evaluate a batch below